File: app/src/main/java/com/singword/app/data/remote/LyricsRepository.kt
package com.singword.app.data.remote

class LyricsRepository(
    private val primary: LyricsDataSource,
    private val secondary: LyricsDataSource? = null
) {
    suspend fun searchLyrics(query: String): LyricsResult {
        return when (val primaryResult = primary.search(query)) {
            is LyricsResult.Success -> primaryResult
            is LyricsResult.NotFound -> secondary?.search(query) ?: LyricsResult.NotFound
            is LyricsResult.NetworkError -> primaryResult
            is LyricsResult.ProviderError -> primaryResult
        }
    }

    suspend fun searchCandidates(query: String, limit: Int = 5): LyricsCandidateResult {
        return when (val primaryResult = searchCandidates(primary, query, limit)) {
            is LyricsCandidateResult.Success -> primaryResult
            is LyricsCandidateResult.NotFound -> {
                val backup = secondary ?: return LyricsCandidateResult.NotFound
                searchCandidates(backup, query, limit)
            }
            is LyricsCandidateResult.NetworkError -> primaryResult
            is LyricsCandidateResult.ProviderError -> primaryResult
        }
    }

    private suspend fun searchCandidates(
        dataSource: LyricsDataSource,
        query: String,
        limit: Int
    ): LyricsCandidateResult {
        val candidateSource = dataSource as? LyricsCandidateDataSource
        if (candidateSource != null) {
            return candidateSource.searchCandidates(query, limit)
        }

        return when (val result = dataSource.search(query)) {
            is LyricsResult.Success -> LyricsCandidateResult.Success(
                candidates = listOf(
                    LyricsCandidate(
                        trackName = result.trackName,
                        artistName = result.artistName,
                        lyrics = result.lyrics,
                        provider = result.provider
                    )
                ),
                provider = result.provider
            )
            is LyricsResult.NotFound -> LyricsCandidateResult.NotFound
            is LyricsResult.NetworkError -> LyricsCandidateResult.NetworkError(result.message)
            is LyricsResult.ProviderError -> LyricsCandidateResult.ProviderError(
                provider = result.provider,
                message = result.message
            )
        }
    }
}

File: ios/singword/Networking/LyricsRepository.swift
import Foundation

final class LyricsRepository {
    private let primary: LyricsDataSource
    private let secondary: LyricsDataSource?

    init(primary: LyricsDataSource, secondary: LyricsDataSource? = nil) {
        self.primary = primary
        self.secondary = secondary
    }

    func searchLyrics(_ query: String) async -> LyricsResult {
        switch await primary.search(query: query) {
        case .success(let trackName, let artistName, let lyrics, let provider):
            return .success(trackName: trackName, artistName: artistName, lyrics: lyrics, provider: provider)
        case .notFound:
            guard let secondary else { return .notFound }
            return await secondary.search(query: query)
        case .networkError(let message):
            return .networkError(message)
        case .providerError(let provider, let message):
            return .providerError(provider: provider, message: message)
        }
    }

    func searchCandidates(_ query: String, limit: Int = 5) async -> LyricsCandidateResult {
        switch await searchCandidates(using: primary, query: query, limit: limit) {
        case .success(let candidates, let provider):
            return .success(candidates: candidates, provider: provider)
        case .notFound:
            guard let secondary else { return .notFound }
            return await searchCandidates(using: secondary, query: query, limit: limit)
        case .networkError(let message):
            return .networkError(message)
        case .providerError(let provider, let message):
            return .providerError(provider: provider, message: message)
        }
    }

    private func searchCandidates(
        using dataSource: LyricsDataSource,
        query: String,
        limit: Int
    ) async -> LyricsCandidateResult {
        if let candidateSource = dataSource as? LyricsCandidateDataSource {
            return await candidateSource.searchCandidates(query: query, limit: limit)
        }

        switch await dataSource.search(query: query) {
        case .success(let trackName, let artistName, let lyrics, let provider):
            return .success(
                candidates: [
                    LyricsCandidate(
                        trackName: trackName,
                        artistName: artistName,
                        lyrics: lyrics,
                        provider: provider,
                        duration: nil
                    )
                ],
                provider: provider
            )
        case .notFound:
            return .notFound
        case .networkError(let message):
            return .networkError(message)
        case .providerError(let provider, let message):
            return .providerError(provider: provider, message: message)
        }
    }
}
